### PR TITLE
Replace InstrumentedFunction by ExperimentFunction (different API)

### PR DESCRIPTION
**Pull request by jrapin**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Improve reproducibility of random functions, simplifies API, solves bugs

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by jrapin** [comment]
@teytaud this solves the reproducibility issues
Eventually (= as soon as I have time to update the code) `InstrumentedFunction` will disappear in favor of `ExperimentFunction`.
The main differences are:
- the initialization takes 1 function and 1 instrumentation (instead of 1 function and possibly many variables)
- `__call__` directly forwards the arguments to the function, without converting to data space (while before it would take a `np.ndarray` from the data space
- it has a `copy` method, which makes sure we can a new version with new random values we can seed (this is what makes everything work)

**Comment by jrapin** [comment]
TODO: update doc

File: nevergrad/benchmark/xpbase.py [hunk 1]
```diff
@@ -8,42 +8,20 @@
...
+from ..functions.rl.agents import torch  # import includes pytorch fix
+from ..functions import base as fbase
+from ..optimization import base as obase
 from ..optimization.optimizerlib import registry as optimizer_registry  # import from optimizerlib so as to fill it
 from . import execution
 
 registry = decorators.Registry[Callable[..., Iterator['Experiment']]]()
 
 
-class IFuncWrapper(execution.PostponedObject):
```
Comment (jrapin) [inline nevergrad/benchmark/xpbase.py]: No more need for this wrapper, which is good :)

File: nevergrad/benchmark/xpbase.py [hunk 2]
```diff
@@ -139,22 +116,20 @@ class Experiment:
...
                  batch_mode: bool = True, seed: Optional[int] = None,
                  cheap_constraint_checker: Optional[Callable[[Any], Any]] = None,
                  ) -> None:
-        assert isinstance(function, instru.InstrumentedFunction), ("All experiment functions should derive from InstrumentedFunction")
+        assert isinstance(function, fbase.ExperimentFunction), ("All experiment functions should "
+                                                                "derive from ng.functions.ExperimentFunction")
         self.function = function
         # Conjecture on the noise level.
-        if not self.function.instrumentation.probably_noisy:
-            if hasattr(self.function, "_parameters") and "noise_level" in self.function._parameters and self.function._parameters["noise_level"] > 0:  # type: ignore
```
Comment (jrapin) [inline nevergrad/benchmark/xpbase.py]: @teytaud the `probably_noisy` tag is now dealt with directly in `ArtificialFunction`, which is safer

File: nevergrad/functions/photonics/core.py [hunk 3]
```diff
@@ -124,7 +125,8 @@ def __init__(self, name: str, dimension: int, transform: str = "tanh") -> None:
         self._func = inst.CommandFunction(["octave-cli", "--no-gui", "--no-history", "--norc", "--quiet", "--no-window-system", path.name],
                                           cwd=path.parent, verbose=False,
                                           env=dict(os.environ, OMP_NUM_THREADS="1", OPENBLAS_NUM_THREADS="1"))
-        super().__init__(self._compute, *_make_instrumentation(name=name, dimension=dimension, transform=transform).args)
+        super().__init__(self._compute, _make_instrumentation(name=name, dimension=dimension, transform=transform))
+        self.register_initialization(name=name, dimension=dimension, transform=transform)
```
Comment (jrapin) [inline nevergrad/functions/photonics/core.py]: just to make it clear once: `register_initialization` keeps a record of the initialization parameters so that the "copy" method can reinstantiate a new object easily

File: nevergrad/functions/rl/agents.py [hunk 4]
```diff
@@ -129,10 +129,11 @@ def __init__(
...
         assert isinstance(env_runner.env, gym.Env)
         self.agent = agent.copy()
-        self.runner = env_runner
+        self.runner = env_runner.copy()
         self.reward_postprocessing = reward_postprocessing
-        super().__init__(self.compute, **self.agent.instrumentation.kwargs)
-        self._descriptors.update(num_repetitions=self.runner.num_repetitions, instrumentation="")
+        super().__init__(self.compute, self.agent.instrumentation.copy().with_name(""))
+        self.register_initialization(agent=agent, env_runner=env_runner, reward_postprocessing=reward_postprocessing)
+        self._descriptors.update(num_repetitions=self.runner.num_repetitions, archi=self.agent.module.__class__.__name__)
```
Comment (jrapin) [inline nevergrad/functions/rl/agents.py]: I've added the `archi` descriptor here. It takes the module class name which is `Perceptron` or `DenseNet`. This will replace `mono` and `multi`  which were manually added in the experiments (which does not work anymore -> descriptors should preferably be automatic to avoid errors)

File: nevergrad/functions/stsp/core.py [hunk 5]
```diff
@@ -11,17 +11,18 @@
...
-    def __init__(self, seed: int = 0, dimension: int = 500) -> None:
+    def __init__(self, dimension: int = 500) -> None:
         super().__init__(self._simulate_stsp, inst.var.Array(dimension))
-        self.order = np.arange(0, self.instrumentation.dimension)
-        self.instrumentation.random_state.seed(seed)
-        self.x = self.instrumentation.random_state.normal(size=self.dimension)
-        self.y = self.instrumentation.random_state.normal(size=self.dimension)
+        self.register_initialization(dimension=dimension)
+        self.order = np.arange(0, self.dimension)
+        self.x = self.parametrization.random_state.normal(size=self.dimension)
```
Comment (jrapin) [inline nevergrad/functions/stsp/core.py]: just to make it clear once, `instrumentation` becomes `parametrization` in `ExperimentFunction` for forward compatibility with future updates (see #391 ), but the underlying instance is still an `Instrumentation`. At some point in the near future, this will also be the case for optimizers (but after the merge of new parametrization framework).

File: nevergrad/functions/utils.py [hunk 6]
```diff
@@ -27,30 +26,3 @@ def __call__(self, x: np.ndarray) -> np.ndarray:
         if self.rotation_matrix is not None:
             y = self.rotation_matrix.dot(y)
         return y
-
-
-class PostponedObject(abc.ABC):
```
Comment (jrapin) [inline nevergrad/functions/utils.py]: Both those classes are not necessary anymore since the methods are directly included in `ExperimentFunction`

**Comment by jrapin** [comment]
Reviewed with @teytaud, ready to go
